compiler/natives/src/net/http/fetch.go

```diff
@@ -0,0 +1,143 @@
+// +build js
+
+package http
+
+import (
+	"errors"
+	"io"
+	"io/ioutil"
+	"strconv"
+
+	"github.com/gopherjs/gopherjs/js"
+)
+
+// streamReader implements an io.ReadCloser wrapper for ReadableStream of https://fetch.spec.whatwg.org/.
+type streamReader struct {
+	pending []byte
+	stream  *js.Object
+}
+
+func (r *streamReader) Read(p []byte) (n int, err error) {
+	if len(r.pending) == 0 {
+		var (
+			bCh   = make(chan []byte)
+			errCh = make(chan error)
+		)
+		r.stream.Call("read").Call("then",
+			func(result *js.Object) {
+				if result.Get("done").Bool() {
+					errCh <- io.EOF
+					return
+				}
+				bCh <- result.Get("value").Interface().([]byte)
+			},
+			func(reason *js.Object) {
+				// Assumes it's a DOMException.
+				errCh <- errors.New(reason.Get("message").String())
+			},
+		)
+		select {
+		case b := <-bCh:
+			r.pending = b
+		case err := <-errCh:
+			return 0, err
+		}
+	}
+	n = copy(p, r.pending)
+	r.pending = r.pending[n:]
+	return n, nil
+}
+
+func (r *streamReader) Close() error {
+	// TOOD: Cannot do this because it's a blocking call, and Close() is often called
+	//       via `defer resp.Body.Close()`, but GopherJS currently has an issue with supporting that.
+	//       See https://github.com/gopherjs/gopherjs/issues/381 and https://github.com/gopherjs/gopherjs/issues/426.
+	/*ch := make(chan error)
+	r.stream.Call("cancel").Call("then",
+		func(result *js.Object) {
+			if result != js.Undefined {
+				ch <- errors.New(result.String()) // TODO: Verify this works, it probably doesn't and should be rewritten as result.Get("message").String() or something.
+				return
+			}
+			ch <- nil
+		},
+	)
+	return <-ch*/
+	r.stream.Call("cancel")
+	return nil
+}
+
+// fetchTransport is a RoundTripper that is implemented using Fetch API. It supports streaming
+// response bodies.
+type fetchTransport struct{}
+
+func (t *fetchTransport) RoundTrip(req *Request) (*Response, error) {
+	headers := js.Global.Get("Headers").New()
+	for key, values := range req.Header {
+		for _, value := range values {
+			headers.Call("append", key, value)
+		}
+	}
+	opt := map[string]interface{}{
+		"method":  req.Method,
+		"headers": headers,
+	}
+	if req.Body != nil {
+		// TODO: Find out if request body can be streamed into the fetch request rather than in advance here.
+		//       See BufferSource at https://fetch.spec.whatwg.org/#body-mixin.
+		body, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			req.Body.Close() // RoundTrip must always close the body, including on errors.
```

Comment:
Yes, but I don't think it'd be cleaner. You still have to check that `req.Body` is not `nil` before doing it.

There are no other returns before this section of code (the code above simply deals with setting up the fields needed for the request), so `req.Body.Close()` will always be called.